### DISCUSSION TITLE
왜 JpaRepository를 impelement한 인터페이스를 사용하기만 하는데, save()와 같은 함수를 통해 DB에 저장할 수 있나?

### DISCUSSION THREAD

**Discussion by nayonsoso** [2024-woowacourse-study/level-interview#130] [Spring Data]
JPA 인터페이스 상속 관계에 대해서 언급 해주면 좋을듯요~  SimpleJpaRepository 등

**Comment by jimi567** [comment]
Spring Data JPA 에서 제공하는 인터페이스를 확장하여 사용하므로. 간단한 CRUD 메소드를 사용할 수 있게 됩니다.

### Spring Data JPA 동작원리

1. 인터페이스 정의 

개발자가 자신의 리포지토리 클래스를 JpaRepository와 같은 Spring Data JPA에서 제공하는 인터페이스를 상속하여 구현합니다  

2. 런타임 자동 생성  

런타임 시점에 인터페이스의 구현체를 자동으로 생성하게 됩니다.  
이때, 기본적인 JpaRepository를 구현하게 된다면, SimpleJpaRepository라는 기본 구현 클래스를 동적으로 생성하게 됩니다.  
추가적으로, Spring은  @EnableJpaRepositories 어노테이션을 사용하여 리포지토리 인터페이스를 검색하고, 이를 기반으로 프록시 객체를 생성합니다.  

3. 엔티티 매니저에 위임  

생성된 프록시 객체가 리포지토리 인터페이스의 메소드 호출을 엔티티 매니저에 위임하게 됩니다.  
save가 호출되면, 프록시 객체는 이 메소드를 엔티티 매니저의 persist 메소드(엔티티가 비영속 상태면), merger(엔티티가 준영속 상태) 등을 호출하여 데이터를 저장하게 됩니다. 

---
추가적으로  
JpaRepository 인터페이스는 save 이외에도   
findAll, saveAll, delete, findById 등의 기본적인 CRUD 기능을 제공합니다  그 외에도 JpaRepository는 다양한 인터페이스를 상속 받고 있기 때문에 페이징이나 정렬 기능 등도 확장하여 사용할 수 있습니다.  
![image](https://github.com/2024-woowacourse-study/level-interview/assets/28584160/aa4aa776-eb77-4001-a9c4-09c8ec700a4d)

**Comment by GIVEN53** [comment]
### 동작 과정
1. `JpaRepository`를 확장한 MemberRepository를 구현합니다.
2. 애플리케이션이 시작될 때(런타임 시) 스프링 부트는 `@EnableJpaRepositories` 어노테이션을 통해 Spring Data JPA Repository 인터페이스를 스캔합니다.
3. Spring Data JPA는 MemberRepository 인터페이스에 대해 SimpleJpaRepository를 포함하는 프록시 객체를 생성합니다.
4. 생성된 프록시 객체는 Bean으로 등록됩니다.
5. SimpleJpaRepository에 `save()`와 같은 메서드들이 오버라이딩되어 있어 `save()` 호출 시 `em.save()`가 동작합니다.
